### PR TITLE
Spectrum grouping

### PR DESCRIPTION
**Pull request by JouvinLea**
Grouping coding test

### CODE REVIEW

**Comment by joleroi** [comment]
This look good!

File: docs/development/howto.rst [hunk 1]
```diff
@@ -11,6 +11,12 @@ in the form of short "How to" or "Q & A" entries.
 
 .. _make_clean:
 
+How to get a development version of the code
+--------------------------------------------
+TODO: 
+
```
Comment (joleroi) [inline docs/development/howto.rst]: I would remove the ":"

File: gammapy/spectrum/counts_spectrum.py [hunk 2]
```diff
@@ -58,6 +57,14 @@ def __init__(self, counts, energy, livetime=None, backscal=1):
             self._livetime = livetime
         else:
             self._livetime = Quantity(0, 's')
+        if muoneff is not None:
+            self.muoneff = muoneff
+        else:
+            self.muoneff = 0
```
Comment (joleroi) [inline gammapy/spectrum/counts_spectrum.py]: I would not set the default value to 0 but rather leave the attribute as 'None'. Here and also for the zenith angle the value 0 hast a physical meaning but does not make sense as a default value. The same hold for the livetime actually. An elegant way to write this is

```
self._livetime = None if livetime is None else Quantity(livetime)
```

**Comment by joleroi** [comment]
@JouvinLea : Following our discussion this morning. I think we agreed that some of the functionality that you introduce in this PR, should be added to existing classes/functions in gammapy. Since it is a lot of functionality this is not trivial. I suggest we go step-by-step:

* Start a new branch from the master and copy over the code from this PR to the right places
```
git checkout master
git pull
git checkout -b grouping_v2
```
* Add the functionality to stack IRFs

I have done this for the counts spectrum (aka PHA) here
https://github.com/gammapy/gammapy/blob/master/gammapy/spectrum/counts_spectrum.py#L80
For me this way of adding two counts spectra is very intuitive
```
from gammapy.utils.energy import EnergyBounds
from gammapy.spectrum import CountsSpectrum

ebounds = EnergyBounds.equal_log_spacing(1,10,10,'TeV')
counts1 = [6,3,8,4,9,5,9,5,5,1]
counts2 = [5,3,2,4,4,5,2,1,8,4]
counts_spec1 = CountsSpectrum(counts1, ebounds)
counts_spec2 = CountsSpectrum(counts2, ebounds)

counts_spec_sum = counts_spec1 + counts_spec2
```
or in the case of many spectra
```
counts_spec_sum = np.sum([counts_spec1, counts_spec2])
```
But I am of course open for other suggestions


It is now quite easy to test the functionality
```
a = counts_spec1.counts[5]
b = counts_spec2.counts[5]
c = counts_spec_sum.counts[5]
a + b == c
Out[17]: True
```
Excatly this test is implemented here
https://github.com/gammapy/gammapy/blob/master/gammapy/spectrum/tests/test_counts_spectrum.py#L45

* Add grouping code: I would leave this as next step once this above has been achieved (I am not good a thinking all at once, I have to do everything in small steps)

-----------------

To sum up: In my opinion it would be best to start be adding similar function to the ARF and RMF classes
https://github.com/gammapy/gammapy/blob/master/gammapy/irf/energy_dispersion.py#L18
https://github.com/gammapy/gammapy/blob/master/gammapy/irf/effective_area_table.py#L67
This way you can use it later for the grouping code in a very convenient way

**Comment by cdeil** [comment]
Completely agree with @joleroi that combining observations is the first step and it would be good to add that functionality first and one PR, and to leave the full spectral analysis to a future PR.

I'm not a huge fan of overloading the `+` operator (by defining `__add__`) to combine observations.
For counts it's OK because the operation really is add, and `(a+b)+c=a+(b+c)`
For exposures and energy dispersions the operation is not adding, and I'm not even sure it works.

The best reference I know for combining spectra is CIAO:
* http://cxc.harvard.edu/ciao/ahelp/combine_spectra.html
* http://cxc.harvard.edu/ciao/download/doc/combine.pdf
* http://cxc.harvard.edu/ciao/threads/coadding/

@joleroi - Should we use that directly? Or wrap it? Or re-implement in Gammapy?

If we re-implement, I think this pattern is simple and nice:

```python
class SomeThing:

    @classmethod
    def combine(cls, some_thing_list):
        data = ... # compute combined data from per-obs data
        return cls(data)
```

To use it, as an example:
```python
aeff_per_obs = [EffectiveArea(...), EffectiveArea(...)]
aeff_combined = EffectiveArea.combine(aeff_per_obs)
```
I'm not sure if `combine` is the best word, it could also be "average" or "stack" ...
Also, to compute the combined EDISP one also needs AEFF, right?
So maybe it should be this?
```python
observations = # list of Observation classes that contain AEFF, EDISP per obs
observation = Observation.combine(observations)
```

**Comment by joleroi** [comment]
@cdeil : thanks for your input
I agree that overloading the `` __add__ `` function is not the best way to go.

About reusing the CIAO user-contributed script, I have some concerns
* Caveats (from http://cxc.harvard.edu/ciao/ahelp/combine_spectra.html)
    * Combining background spectra with wildly varying spectral extraction region areas may yield misleading uncertainty estimates; i.e., some extractions will be over-represented while others will be under-represented. 
    * Non-chandra files may not work correctly. 
* Code quality (we might have to debug/change the script)
    * https://gist.github.com/joleroi/138fef79ea659097c274
    * 1745 lines of medium-quality code
    * Essential functionality no super-hard to implement
    * We want to get away from scripts that are basically sent via mail
* Output files
    * The output files will be standard PHA files, no ``THRES_LO`` etc. keywords that we wanted to add

So, bottom line: I would reimplement this (copying over the bits and pieces we can use). What do you think @JouvinLea, in the end it is up to you if you want to re-implement stuff or write wrapper functions.

About the API:
Since we need the RMF and ARF file for this function (I did not know that) the RMF and ARF classes are probably not a convenient place for this. Existing class that know about the OGIP files are the ``SpectrumAnalysis`` and ``SpectralFit``, so the method @cdeil  suggested would be added there. 

I would suggest to add a plain function
```
def combine_observations([CountsSpectrum1, ...], [EnergyDispersion,1 ...], [EffectiveAreaTable, 1...])
```
That takes 3 lists of gammapy objects and returns 3 gammapy objects. This can the be used
* by the SpectrumAnalysis class on the Gammapy objects
* by some command line tool on the PHA files
* by the SpectralFit class on the PHA files

This way we stay flexible until we have a better understanding of where this grouping makes the most sense.

**Comment by cdeil** [comment]
@joleroi - Implementing the obs stacking in Gammapy also sounds like a good solution to me (without having read through the CIAO combine_spectra code).

In terms of API: a standalone function is probably OK.

Or maybe the [SpectrumObservation](https://gammapy.readthedocs.org/en/latest/api/gammapy.spectrum.SpectrumObservation.html) class?
The docstring doesn't really say much about it's purpose and content, but it does store the spectrum data members we are talking about:
```
        self._on = None
        self._off = None
        self._aeff = None
        self._edisp = None
```
so the classmethod idea I mentioned above could fit there:
``` 
class SpectrumObservation:
    @classmethod
    def combine(cls, spectrum_observation_list):
```
(In the end the difference between a standalone function and a classmethod is not large, and anyways I don't have a strong opinion on this ...)

**Comment by cdeil** [comment]
Changed milestone to 0.5

**Comment by joleroi** [comment]
@JouvinLea Would you like to reuse this or https://github.com/gammapy/gammapy/pull/429 or should we close these PRs and open a new one?

**Comment by cdeil** [comment]
Superseded by #471.
